### PR TITLE
Feat: FeedSwitch 

### PR DESCRIPTION
**Pull request by 0xtj24**

### CODE REVIEW

File: src/util/FeedSwitch.sol [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        timelockPeriod = _timelockPeriod;
+        maturity = _maturity;
+        guardian = _guardian;
+    }
+
+    /// @notice Initiate the feed switch, entering the timelock period
+    /// @dev Can only be called by the guardian
+    function initiateFeedSwitch() external {
+        if (msg.sender != guardian) revert NotGuardian();
+        switchInitiatedAt = block.timestamp;
```
Comment (08xmt) [inline src/util/FeedSwitch.sol:55]: Avoid giving the guardian the ability to continually bump up the initiation time.
```suggestion
        if(switchInitiatedAt == 0)
            switchInitiatedAt = block.timestamp;
```

File: src/util/FeedSwitch.sol [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        returns (
+            uint80 roundId,
+            int256 price,
+            uint256 startedAt,
+            uint256 updatedAt,
+            uint80 answeredInRound
+        )
+    {
+        return feed.latestRoundData();
+    }
```
Comment (08xmt) [inline src/util/FeedSwitch.sol:96]: I think it makes more sense to switch to the `afterMaturityFeed` as soon as the PT tokens mature, and having the feed automatically roll over to the the `beforeMaturityFeed` as soon as the `timelockPeriod` has elapsed.
```suggestion
    /// @notice Get the current feed data
    /// @return roundId The round ID
    /// @return price The price of the asset
    /// @return startedAt The timestamp of the start of the round
    /// @return updatedAt The timestamp of the last update
    /// @return answeredInRound The round ID in which the price was answered
    function latestRoundData()
        public
        view
        returns (
            uint80 roundId,
            int256 price,
            uint256 startedAt,
            uint256 updatedAt,
            uint80 answeredInRound
        )
    {
        if(block.timestamp > maturity){
            return afterMaturityFeed.latestRoundData();
        } else if (
            switchInitiatedAt > 0 &&
            switchInitiatedAt + timelockPeriod < block.timestamp)
        {
            return beforeMaturityFeed.latestRoundData();
        }
        return feed.latestRoundData();
    }
```

File: src/util/FeedSwitch.sol [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        uint256 _maturity,
+        address _guardian
+    ) {
+        feed = IChainlinkFeed(_feed);
+        beforeMaturityFeed = IChainlinkFeed(_beforeMaturityFeed);
+        afterMaturityFeed = IChainlinkFeed(_afterMaturityFeed);
+        if (
+            feed.decimals() != beforeMaturityFeed.decimals() ||
+            feed.decimals() != afterMaturityFeed.decimals() ||
+            feed.decimals() != 18
```
Comment (08xmt) [inline src/util/FeedSwitch.sol:43]: Slightly simpler and equivalent
```suggestion
            beforeMaturityFeed.decimals() != 18 ||
            afterMaturityFeed.decimals() != 18 ||
            feed.decimals() != 18
```

File: src/util/FeedSwitch.sol [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        ) revert FeedDecimalsMismatch();
+
+        timelockPeriod = _timelockPeriod;
+        maturity = _maturity;
```
Comment (08xmt) [inline src/util/FeedSwitch.sol:47]: Not super important, but consider a constructor check that makes sure that maturity is in the future.

File: src/util/FeedSwitch.sol [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+            uint80 roundId,
+            int256 price,
+            uint256 startedAt,
+            uint256 updatedAt,
+            uint80 answeredInRound
+        )
+    {
+        if (block.timestamp > maturity) {
+            return afterMaturityFeed.latestRoundData();
+        } else if (block.timestamp > switchCompletedAt) {
```
Comment (08xmt) [inline src/util/FeedSwitch.sol:101]: If switch is completed in this block, it should count.
```suggestion
        } else if (block.timestamp >= switchCompletedAt) {
```

File: src/util/FeedSwitch.sol [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+    /// @notice Initiate the feed switch, entering the timelock period
+    /// @dev Can only be called by the guardian and can be done before maturity, after that, the feed will return afterMaturityFeed data
+    function initiateFeedSwitch() external {
+        if (msg.sender != guardian) revert NotGuardian();
+        if (block.timestamp > maturity) revert MaturityPassed();
+
+        if (switchCompletedAt == 0 || switchCompletedAt < block.timestamp) {
+            switchCompletedAt = block.timestamp + timelockPeriod;
+        } else switchCompletedAt = 0;
```
Comment (08xmt) [inline src/util/FeedSwitch.sol:69]:
```suggestion
        if (switchCompletedAt > block.timestamp) revert SwitchInProgress();
        
        switchCompletedAt = block.timestamp + timelockPeriod;
```

File: src/util/FeedSwitch.sol [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        guardian = _guardian;
+    }
+
+    /// @notice Initiate the feed switch, entering the timelock period
+    /// @dev Can only be called by the guardian and can be done before maturity, after that, the feed will return afterMaturityFeed data
+    function initiateFeedSwitch() external {
+        if (msg.sender != guardian) revert NotGuardian();
+        if (block.timestamp > maturity) revert MaturityPassed();
+
+        if (switchCompletedAt == 0 || switchCompletedAt < block.timestamp) {
```
Comment (08xmt) [inline src/util/FeedSwitch.sol:67]: Remove unnecessary check
```suggestion
        if (switchCompletedAt < block.timestamp) {
```

File: src/util/FeedSwitch.sol [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+        guardian = _guardian;
+    }
+
+    /// @notice Initiate the feed switch, entering the timelock period
+    /// @dev Can only be called by the guardian and can be done before maturity, after that, the feed will return afterMaturityFeed data
+    function initiateFeedSwitch() external {
+        if (msg.sender != guardian) revert NotGuardian();
+        if (block.timestamp > maturity) revert MaturityPassed();
+
+        if (switchCompletedAt < block.timestamp) {
```
Comment (wavey0x) [inline src/util/FeedSwitch.sol:67]: suggestion:
from UX pov, not great to have a single 0-parameter function do multiple different things.
1. queue switch
1. cancel switch

could we benefit from a helper funcs like this?:
- `isFeedSwitchQueued() public returns (bool)`

... may not be strictly necessary, but could help guardian know exactly what the consequence of their next call will be.

**Review by wavey0x — APPROVED** [review]

File: src/util/FeedSwitch.sol [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        view
+        returns (
+            uint80 roundId,
+            int256 price,
+            uint256 startedAt,
+            uint256 updatedAt,
+            uint80 answeredInRound
+        )
+    {
+        if (block.timestamp > maturity) {
```
Comment (wavey0x) [inline src/util/FeedSwitch.sol:99]: im not an expert on pendle, but does this match pendle logic? i.e. does maturity effect take place only *after* the timestamp, or on it?

**Review by wavey0x — APPROVED** [review]

**Review by 08xmt — APPROVED** [review]

**Pull request merged**